README.md
# NDBC

![alt text](http://www.ndbc.noaa.gov/images/nws/noaaleft.jpg "NOAA") ![alt text](http://www.ndbc.noaa.gov/images/nws/ndbc_title.jpg "NDBC")

This repository represents my attempts to build out Python class(es)
to facilitate the acquisition, analysis, and visualization of National
Data Buoy Center (NDBC) data. The goal is to develop a set of APIs to
facilitate rapid discovery of data resources, exploratory data analysis,
and allow integration into automated data workflows.

## NDBC.py

This file defines the DataBuoy class. The purpose of this class is to
allow a user to define a specific data buoy they wish to gather data
from and provide the user with methods to collect and analyze this data.

Dependencies are listed in `requirements.txt`

## Usage

#### Installation

Install using pip from PyPI

```
pip install NDBC
```

Then you are ready to start using this module in exploratory data analyses and scripted workflows.

#### Methods of DataBuoy Class

`.set_station_id`

If a DataBuoy class has been instantiated without any `station_id` argument, this method allows for setting a station id

```
from NDBC.NDBC import DataBuoy
DB = DataBuoy()
DB.set_station_id('46042') # <- Either strings or numbers are acceptable
```

`.get_station_metadata()`

Perform a scrape of the public webpage for a specified data station and save a dictionary of available metadata to the `.station_info` property. This is only available if a DataBuoy has a valid `station_id` set (either during class instantiation or using
the `set_station_id` method).

```
from NDBC.NDBC import DataBuoy
DB = DataBuoy(46042)
DB.get_station_metadata()
DB.station_info
{   'Air temp height': '4 m above site elevation',
    'Anemometer height': '5 m above site elevation',
    'Barometer elevation': 'sea level',
    'Sea temp depth': '0.6 m below water line',
    'Site elevation': 'sea level',
    'Watch circle radius': '1789 yards',
    'Water depth': '1645.9 m',
    'lat': '36.785 N',
    'lon': '122.398 W'}
```

- `.get_data(datetime_index=False)`

After importing, the DataBuoy class is instantiated with the ID of the
station from which historical data is sought. Then data may be gathered for
the years and months specified. If no time period is specified, the most recent
full month available is retrieved.

The default behavior is to append datetime values built from date part columns (YY, MM, DD, etc.) to a column 'datetime'. If value `True` is passed as the `datetime_index` argument, the datetime values will be used as index values for the returned dataframe. In some cases this is advantageous for time series analyses.

```
from NDBC.NDBC import DataBuoy

n42 = DataBuoy(46042)  # <- String or numeric station ids are valid

n42.get_data(datetime_index=True)  # <- no year, month argumets so latest full month is retrieved. Default data type is 'stdmet'

Oct not available.   # <- Where data is missing, messages are returned to the terminal via a logger.warning() call
Sep not available.

n42.data  # <- anticipating additional data collection methods, the .data property returns a dictionary.  Indiviudual
               data products are returned as pandas DataFrame objects

# Datetime objects are compiled from individual year, month, day, hour, minute columns and used as the index to support
# slicing data by time frames.

{'stdmet':          WDIR WSPD  GST  WVHT    DPD   APD  MWD    PRES  ATMP  WTMP   DEWP   VIS   TIDE
2019-07-31 23:50:00  298  3.6  5.2  1.25   7.69  5.37  303  1015.1  13.4  15.2  999.0  99.0  99.00
2019-08-01 00:50:00  301  5.7  7.2  1.26   7.14  5.42  306  1014.8  13.4  15.3  999.0  99.0  99.00
2019-08-01 01:50:00  323  6.6  8.3  1.33   7.14  5.47  312  1014.5  13.2  15.1  999.0  99.0  99.00
2019-08-01 02:50:00  347  5.8  7.7  1.32   7.69  5.15  319  1014.5  12.7  15.1  999.0  99.0  99.00
2019-08-01 03:50:00  353  5.6  7.2  1.26   7.69  5.31  325  1014.9  12.6  15.0  999.0  99.0  99.00
...                  ...  ...  ...   ...    ...   ...  ...     ...   ...   ...    ...   ...    ...
2019-08-31 18:50:00  999  6.2  7.4  0.87  13.79  4.67  186  1014.6  17.0  17.2  999.0  99.0  99.00
2019-08-31 19:50:00  999  6.8  8.3  0.83  13.79  4.56  178  1014.2  17.2  17.3  999.0  99.0  99.00
2019-08-31 20:50:00  999  6.5  7.8  0.89  13.79  4.38  195  1013.8  17.5  17.4  999.0  99.0  99.00
2019-08-31 21:50:00  999  7.5  8.9  0.95  13.79  4.52  190  1013.1  17.5  17.3  999.0  99.0  99.00
2019-08-31 22:50:00  999  8.0  9.4  0.95  13.79  4.09  171  1012.7  17.7  17.1  999.0  99.0  99.00

[741 rows x 13 columns]}
```

By default the `get_data()` function will fetch the most current month's data. However, the function can take lists of years & months ([int]) to specify a timeframe.

```
>>> n42 = NDBC.DataBuoy('46042')
>>> n42.get_data(months=[1,2], years=range(2019, 2020), datetime_index=True, data_type='swden)
Year 2019 not available.
Year 2020 not available.
 
>>> n42.data
{'swden': {'data':                      .0200  .0325  .0375  .0425  .0475  .0525  .0575  .0625  .0675  .0725  .0775  .0825  .0875  ...  .3000  .3100  .3200  .3300  .3400  .3500  .3650  .3850  .4050  .4250  .4450  .4650  .4850
2021-01-01 00:40:00    0.0    0.0    0.0   0.00   1.17   9.11  24.25  24.95  15.84  20.44  26.48  20.63  12.72  ...   0.28   0.31   0.19   0.20   0.13   0.07   0.06   0.05   0.03   0.01   0.01   0.00    0.0
2021-01-01 01:40:00    0.0    0.0    0.0   0.00   0.00  13.76  26.55  22.40  24.12  30.09  23.41  15.74  14.95  ...   0.25   0.16   0.12   0.16   0.06   0.16   0.06   0.03   0.05   0.02   0.01   0.00    0.0
2021-01-01 02:40:00    0.0    0.0    0.0   0.00   0.93   4.40  16.03  33.95  41.48  38.02  31.47  18.88  14.59  ...   0.21   0.15   0.18   0.14   0.14   0.10   0.07   0.05   0.03   0.02   0.01   0.00    0.0
2021-01-01 03:40:00    0.0    0.0    0.0   0.07   1.14   6.95  27.94  45.68  41.92  30.11  25.03  19.52  10.93  ...   0.22   0.20   0.16   0.09   0.08   0.15   0.09   0.04   0.02   0.01   0.00   0.01    0.0
2021-01-01 04:40:00    0.0    0.0    0.0   0.00   0.76   3.64  11.23  18.23  29.84  27.19  12.85  11.20   9.77  ...   0.13   0.17   0.14   0.16   0.08   0.08   0.07   0.08   0.05   0.01   0.01   0.00    0.0
...                    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...  ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...
2021-02-28 19:40:00    0.0    0.0    0.0   0.00   0.00   0.00   0.06   0.25   1.42   2.50   9.48  11.48   8.46  ...   0.21   0.13   0.11   0.08   0.10   0.04   0.02   0.02   0.03   0.01   0.00   0.00    0.0
2021-02-28 20:40:00    0.0    0.0    0.0   0.02   0.05   0.08   0.24   1.02   3.97   4.97   4.99   8.31  10.09  ...   0.21   0.07   0.09   0.06   0.05   0.10   0.04   0.03   0.01   0.01   0.00   0.00    0.0
2021-02-28 21:40:00    0.0    0.0    0.0   0.00   0.00   0.15   0.30   0.36   1.63   4.18   6.85   7.82   7.98  ...   0.12   0.11   0.09   0.08   0.04   0.05   0.06   0.02   0.01   0.01   0.00   0.00    0.0
2021-02-28 22:40:00    0.0    0.0    0.0   0.00   0.01   0.09   0.10   0.32   2.84   3.82   3.91   4.92   5.17  ...   0.17   0.09   0.13   0.05   0.05   0.08   0.06   0.03   0.01   0.01   0.00   0.00    0.0
2021-02-28 23:40:00    0.0    0.0    0.0   0.00   0.00   0.00   0.18   0.25   1.78   3.97   5.08   4.98   5.40  ...   0.07   0.10   0.11   0.08   0.08   0.06   0.03   0.02   0.01   0.01   0.00   0.00    0.0

[1413 rows x 47 columns]}}

```

Likely due to my own biases in my research interests, the `get_data()`  function will default to fetching
standard meteorological data.  However, users can specify different data packages like so `get_data(data_type='cwind')`.  To view which data packages
are currently supported examine the `DataBuoy.DATA_PACKAGES` attribute:
```
{'cwind': {'name': 'Continous Wind Data', 'url_char': 'c'},
 'srad': {'name': 'Solar radiation data', 'url_char': 'r'},
 'stdmet': {'name': 'Standard meteoroligcal data', 'url_char': 'h'},
 'swden': {'name': 'Spectral Wave Density data', 'url_char': 'w'},
 'swdir': {'name': 'Spectral wave (alpha1) direction data', 'url_char': 'd'},
 'swdir2': {'name': 'Spectral wave (alpha2) direction data', 'url_char': 'i'},
 'swr1': {'name': 'Spectral wave (r1) direction data', 'url_char': 'j'},
 'swr2': {'name': 'Spectral wave (r2) direction data', 'url_char': 'k'}}

```

Using the pandas DataFrame to store the returned data provides access to the wide array of methods the pandas package
provides.

- `.save(filename(optional))`

Saves an instantiated DataBuoy object as JSON to a file. If `filename` is not specified the file name will follow the
`databuoy_{station_id}.json` convention.

```
db = DataBuoy(46042)
db.save('/path/to/file/my_filename.json')
```

_classmethod_

- `.load(filename)`
  Instantiate a DataBuoy object from a file, generated by the `.save()` method.

```
db = DataBuoy.load('/path/to/file.json')
```


changelog.md
# NDBC Package Change Log

---

### v.1.2.0
 - Added `from .NDBC import DataBuoy` to `__init__.py` to reduce redundant import statements.  DataBuoy class can be imported as `from NDBC import DataBuoy`.
 - Moved package details to `setup.cfg` following [python.org best practices](https://packaging.python.org/tutorials/packaging-projects/)
 

### v1.1.1
- bug(fix) Correct issue with alpha characters in station ID string [#32](https://github.com/GenSci/NDBC/issues/32)
- bug(fix) Prevent continuous year looping when URLs fail.


### v.1.1.0
- Added support for multiple data package retrieval through a more general `get_data()` function.
- Added deprecation flags to `get_stdmet()` function, which is being replaced by more general `get_data()`.
- Improved bad data flag detection and replacement with `np.NAN` values.
- Added property decorated getter functions for all identified data packages


### v.1.0.1
- Modified month URL kwargs to better handle selecting months in the previous year. (fixes issue #20)
- Added `__repr__()` method so users can see the station ID when calling the object in the terminal
- Added demonstration Jupyter Notebook to make package usage clear.


getNDBC_stdMet.py
# -*- coding: utf-8 -*-
"""
Created on Fri May 20 19:55:41 2016
GET NDBC
@author: ryan
"""
import pymysql.cursors
import pandas as pd
import configparser
from datetime import datetime as dt
con=configparser.ConfigParser()
con.read('/home/ryan/Python_Scripts/NDBC/config.ini')
myyears=range(1987,2017) # The last number won't be generated
mymonths=range(1,3) # same thing here
td=dt.today()
this_year=td.year
# getting access to our local db
my_host=con['LOCALDB']['host']
my_port=int(con['LOCALDB']['port'])
my_user=con['LOCALDB']['user']
my_passwd=con['LOCALDB']['passwd']
my_db=con['LOCALDB']['db']
conn=pymysql.connect(host=my_host,port=my_port,user=my_user,passwd=my_passwd,
                   db=my_db)
cur=conn.cursor()
# Now let's get our station ids
sql="SELECT station_id FROM ndbc_stations;"
cur.execute(sql)
stations=cur.fetchone()
# Building in fucntionality to replace old names with the correct ones (as of 2014)
def NDBCNames(oldname):
    namedict={  #old    #new
                'YYYY':'YY',
                '#YY':'YY',
                'WD':'WDIR',
                'DIR':'WDIR',
                'SPD':'WSPD',
                'GSP':'GST',
                'GNM':'GTIME',
                'BAR':'PRES',
                'BARO':'PRES',
                'H0':'WVHT',
                'DOMPD':'DPD',
                'AVP':'APD',
                'SRAD':'SWRAD',
                'SRAD2':'SWRAD',
                'LRAD':'LWRAD',
                'LRAD1':'LWRAD'
                }
    if oldname in namedict:
        return namedict[oldname]
    else:
        return oldname

def getData(cursor,station,year,month):
    base_url = 'http://www.ndbc.noaa.gov/view_text_file.php?filename={}'.format(station)
    if month==0:
        target=base_url+'h'+str(year)+'.txt.gz&dir=data/historical/stdmet/'
    elif year==0:
        tm=dt(this_year,month,1)
        target=base_url+str(month)+str(this_year)+'.txt.gz&dir=data/stdmet/'+tm.strftime('%b')+'/'
    else:
        print('Not sure what you meant to do here but this is weird')
    # okay now that we have built our url that should point toward the text data file
    # let's grab that data file

    # we are trainwrecking in 2007 when (at least for station 46042)
    # NOAA started adding a second line to the header with the units
    # for each measurement.  It would be nice to capture this somewhere
    # in our DB but obviously it cannot be part of the numerical data.
    # Perhaps an additional table?
    try:
        my_d=pd.read_csv(target,header=0, engine='python',sep="\s+")
    except:
        if month==0:
            print("Year " +  str(year) + " is not available")
        elif year==0:
            print("Month " + str(month) + " is not available")
        return pd.DataFrame() # in order to properly evaluate the result we need to return an empty data frame...kind of wasteful memory wise but it works
    # Now let's get those pesky year, month, day, and
    # hour columns into a Timestamp
    # first we get the columns with our datetime info
    if my_d.iloc[0][0]=="#yr":
        my_d=insertUnits(cursor,my_d)
    cols=list(my_d.columns)
    if 'mm' in cols:
        cols=cols[0:cols.index('mm')+1]
        my_d.set_index(keys=cols,inplace=True)
    else:
        cols=cols[0:cols.index('hh')+1]
        my_d.set_index(keys=cols,inplace=True)
    my_dt=my_d.index.values
    dtime=[]
    for i in my_dt:
        if len(str(i[0]))==2:
            Y=i[0]+1900
        else:
            Y=i[0]
        if len(cols)==5:
            strtime=dt(int(Y),int(i[1]),int(i[2]),int(i[3]),int(i[4])).isoformat()
        else:
            strtime=dt(Y,i[1],i[2],i[3]).isoformat()
        dtime.append(strtime)
    my_d.index=dtime
    # Huzzah, we have our datetime strings as our index!  Since
    # it is ISO 8601 compliant this should be easily loaded into
    # a MySQL DB.  Now let's handle those bad data flags
    cols=list(my_d.columns) # we need to redo this to account for indexing
    # AFTER CAREFUL CONSIDERATION IT APPEARS TO ME THAT DEALING WITH BAD DATA
    # FLAGS WOULD BE EASIER DONE ONCE THE DATA HAS BEEN LOADED INTO THE DATABASE
    # THEREFORE I WILL SKIP THIS SECTION AND IMPLEMENT A BAD DATA FLAG PROTOCOL
    # LATER ON, AFTER INSERTING THE RECORDS.
    newcols=[];
    for col in cols:
        newcols.append(NDBCNames(col))
#        badFlag=True
#        if my_d[col].max()!=np.nan:
#            flagstr=str(int(my_d[col].max()))
#            flagval=my_d[col].max()
#            for ch in flagstr:
#                if int(ch) != 9:
#                    badFlag=False
#                    break
#            if badFlag:
#                # THIS IS SETTING ALL COLUMNS TO NAN!!!!!
#                # NEED TO FIX THIS
#                my_d[my_d[col]==flagval]=np.NaN
    my_d.columns=newcols
    my_d.fillna(999.0, inplace=True)
    return my_d

def insertStdMet(station,cursor,data):
    # This function takes in a station id, database cursor and an array of data.  At present
    # it assumes the data is a pandas dataaframe with the datetime value as the index
    # It may eventually be modified to be more flexible.  With the parameters
    # passed in, it goes row by row and builds an INSERT INTO SQL statement
    # that assumes each row in the data array represents a new record to be
    # added.
    fields=list(data.columns) # if our table has been constructed properly, these column names should map to the fields in the data table
    # Building the SQL string
    strSQL1='REPLACE INTO std_met (station_id,date_time,'
    strSQL2='VALUES (%s,%s,'
    for f in fields:
        strSQL1+=f+','
        strSQL2+='%s,'
    # trimming the last comma
    strSQL1=strSQL1[:-1]
    strSQL2=strSQL2[:-1]
    strSQL1+=") " + strSQL2 + ")"
    # Okay, now we have our SQL string.  Now we need to build the list of tuples
    # that will be passed along with it to the .executemany() function.
    tuplist=[]
    for i in range(len(data)):
        r=data.iloc[i][:]
        datatup=(station,r.name)
        for f in r:
            datatup+=(f,)
        tuplist.append(datatup)
    cursor.executemany(strSQL1,tuplist)
    conn.commit()
def insertUnits(cursor,data):
    # This function is designed to check whether or not the first line of the
    # pandas data frame passed in contains additional string information, usually
    # the units for the parameters measured.  If so it will insert those units
    # into the units table in our database
    r=data.iloc[0][:]
    if isinstance(r[0],str): # if we have a string we assume r represents the units
        cols=list(data.columns) # getting our list of columns
        strSQL="REPLACE INTO units (parameter, unit) VALUES (%s,%s)"
        tuplist=[]
        if 'mm' in cols:
            start_index=cols.index('mm')+1
        else:
            start_index=cols.index('hh')+1
        for col in cols[start_index:]:
            datatup=(col,r[col])
            tuplist.append(datatup)
        cursor.executemany(strSQL,tuplist)
        data = data.iloc[1:][:]
    return data

# Rolling up both functions into a single function that can be executed using
# a multi-threading approach--not currently using this approach
def wholeShebang(station,year,month,cursor):
    stdmet=getData(station,year,month)
    insertStdMet(station,cursor,stdmet)

threads=[]
for station in stations:
#    for y in myyears:o
#        stdmet=getData(cur,station,y,0)
#        insertStdMet(station,cur,stdmet)
    for m in mymonths:
        stdmet=getData(cur,station,0,m)
        if not stdmet.empty:
            insertStdMet(station,cur,stdmet)
print('Hooray!  You just loaded a crap ton of NDBC data!')
